File: MonityApp/Models/RecurringRule.swift
import Foundation

struct RecurringRule: Codable, Identifiable {
    let id: String
    var amount: Double
    var currency: String
    var type: Transaction.TransactionType
    var frequency: Frequency
    var startDate: String
    var endDate: String?
    var categoryId: Int?
    var category: TransactionCategory?
    var note: String
    var isActive: Bool

    enum Frequency: String, Codable, CaseIterable {
        case daily
        case weekly
        case monthly
        case yearly
    }

    enum CodingKeys: String, CodingKey {
        case id, amount, currency, type, frequency, startDate, endDate
        case categoryId, note, isActive
        case category = "Category"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        amount = try DecodingHelpers.decodeFlexibleDouble(c, forKey: .amount)
        currency = try c.decodeIfPresent(String.self, forKey: .currency) ?? "ILS"
        type = try c.decode(Transaction.TransactionType.self, forKey: .type)
        frequency = try c.decode(Frequency.self, forKey: .frequency)
        startDate = try c.decode(String.self, forKey: .startDate)
        endDate = try c.decodeIfPresent(String.self, forKey: .endDate)
        categoryId = try c.decodeIfPresent(Int.self, forKey: .categoryId)
        category = try c.decodeIfPresent(TransactionCategory.self, forKey: .category)
        note = try c.decodeIfPresent(String.self, forKey: .note) ?? ""
        isActive = try c.decodeIfPresent(Bool.self, forKey: .isActive) ?? true
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(amount, forKey: .amount)
        try c.encode(currency, forKey: .currency)
        try c.encode(type, forKey: .type)
        try c.encode(frequency, forKey: .frequency)
        try c.encode(startDate, forKey: .startDate)
        try c.encodeIfPresent(endDate, forKey: .endDate)
        try c.encodeIfPresent(categoryId, forKey: .categoryId)
        try c.encodeIfPresent(category, forKey: .category)
        try c.encode(note, forKey: .note)
        try c.encode(isActive, forKey: .isActive)
    }
}

struct RecurringListResponse: Codable {
    let recurringRules: [RecurringRule]
}

struct RecurringSingleResponse: Codable {
    let recurringRule: RecurringRule
}

File: backend/src/models/RecurringRule.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const RecurringRule = sequelize.define('RecurringRule', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  amount: {
    type: DataTypes.DECIMAL(12, 2),
    allowNull: false,
  },
  currency: {
    type: DataTypes.STRING(3),
    defaultValue: 'ILS',
  },
  type: {
    type: DataTypes.ENUM('expense', 'income'),
    allowNull: false,
  },
  frequency: {
    type: DataTypes.ENUM('daily', 'weekly', 'monthly', 'yearly'),
    allowNull: false,
  },
  startDate: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  endDate: {
    type: DataTypes.DATEONLY,
    allowNull: true,
  },
  note: {
    type: DataTypes.STRING(500),
    defaultValue: '',
  },
  lastGenerated: {
    type: DataTypes.DATEONLY,
    allowNull: true,
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true,
  },
});

module.exports = RecurringRule;
